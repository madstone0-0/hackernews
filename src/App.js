/* eslint-disable no-unused-vars */
/* eslint-disable no-unused-expressions */
import React, { Component } from "react";
import { StyleSheet, css } from "aphrodite";

const DEFAULT_PAGE = 0;
const DEFAULT_QUERY = "redux";
const DEFAULT_HPP = 100;

const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

/*
 *  Checks whether the term being searched is similar to any
 *  titles.
 */
function isSearched(searchTerm) {
    return function(item) {
        return (
            !searchTerm ||
            item.title.toLowerCase().includes(searchTerm.toLowerCase())
        );
    };
}

const stylesheet = StyleSheet.create({
    page: {
        margin: "20px"
    },

    interactions: {
        textAlign: "center"
    },

    table: {
        margin: "20px 0"
    },

    tableHeader: {
        display: "flex",
        lineHeight: "24px",
        fontSize: "16px",
        padding: "0 10px",
        justifyContent: "space-between",
        span: {
            overflow: "hidden",
            textOverflow: "ellipsis",
            padding: "0 5px"
        }
    },

    tableEmpty: {
        margin: "200px",
        textAlign: "center",
        fontSize: "16px"
    },

    tableRow: {
        display: "flex",
        lineHeight: "24px",
        whiteSpace: "nowrap",
        margin: "10px 0",
        padding: "10px",
        background: "#ffffff",
        border: "1px solid #e3e3e3",
        span: {
            overflow: "hidden",
            textOverflow: "ellipsis",
            padding: "0 5px"
        }
    },

    buttonInline: {
        borderWidth: "0",
        background: "transparent",
        color: "inherit",
        textAlign: "inherit",
        webkitFontSmoothing: "inherit",
        padding: "0",
        fontSize: "inherit",
        cursor: "pointer"
    },

    buttonActive: {
        borderRadius: "0",
        borderBottom: "1px solid #38bb6c"
    }
});

class App extends Component {
    constructor(props) {
        super(props);

        // Sets state
        this.state = {
            result: null,
            searchTerm: DEFAULT_QUERY
        };

        // Declare functions for global use with the this identifier
        this.setSearchTopstories = this.setSearchTopstories.bind(this);
        this.fetchSearchTopstories = this.fetchSearchTopstories.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
        this.onDismiss = this.onDismiss.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
    }

    onSearchSubmit(event) {
        const { searchTerm } = this.state;
        this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
        event.preventDeafult();
    }

    // Set top stories
    setSearchTopstories(result) {
        const { hits, page } = result;
        const oldHits = page !== 0 ? this.state.result.hits : [];
        const updatedHits = [oldHits, hits];

        this.setState({ result: { hits: updatedHits, page } });
    }

    // Fetch the top stories
    fetchSearchTopstories(searchTerm, page) {
        fetch(
            `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}\
${page}&${PARAM_HPP}${DEFAULT_HPP}`
        )
            .then(response => response.json())
            .then(result => this.setSearchTopstories(result));
    }

    // Dismiss stuff
    onDismiss(id) {
        const isNotID = item => item.objectID !== id;
        const updatedHits = this.state.result.hits.filter(isNotID);
        this.setState({
            result: (this.state.result, { hits: updatedHits })
        });
    }

    //Search stuff
    onSearchChange(event) {
        this.setState({ searchTerm: event.target.value });
    }

    render() {
        const { searchTerm, result } = this.state;
        const page = (result && result.page) || 0;
        if (!result) {
            return null;
        }
        return (
            <div className={css(stylesheet.page)}>
                <div className={css(stylesheet.interactions)}>
                    {/* Passed values form the App component to the Search an table components */}
                    <Search
                        value={searchTerm}
                        onChange={this.onSearchChange}
                        onSubmit={this.onSearchSubmit}
                    >
                        Search
                    </Search>
                    <Button
                        onClick={() =>
                            this.fetchSearchTopstories(searchTerm, page + 1)
                        }
                    >
                        More
                    </Button>
                </div>
                {result && (
                    <Table list={result.hits} Dismiss={this.onDismiss} />
                )}
            </div>
        );
    }

    componentDidMount() {
        const { searchTerm } = this.state;
        this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
    }
}

// Functional stateless component Search to handle the search function
const Search = ({ value, onChange, children, onSubmit }) => {
    <form onSubmit={onSubmit}>
        <input type="text" value={value} onChange={onChange} />
        <button type="submit">{children}</button>
    </form>;
};

const largeColumn = {
    width: "40%"
};
const midColumn = {
    width: "30%"
};
const smallColumn = {
    width: "10%"
};

// Functional stateless component Table to handle the appearance of the table
const Table = ({ list, onDismiss }) => {
    <div className={css(stylesheet.table)}>
        {list.map(item => (
            <div key={item.objectID} className={css(stylesheet.tableRow)}>
                <span style={largeColumn}>
                    <a href={item.url}>{item.title}</a>
                </span>
                <span style={midColumn}>{item.author}</span>
                <span style={smallColumn}>{item.num_comments} comments</span>
                <span style={smallColumn}>{item.points} points</span>
                <span style={smallColumn}>
                    <Button
                        onClick={() => onDismiss(item.objectID)}
                        className={css(stylesheet.buttonInline)}
                    >
                        Dismiss
                    </Button>
                </span>
            </div>
        ))}
    </div>;
};

const Button = ({ onClick, className = "", children }) => {
    <button type="button" onClick={onClick} className={className}>
        {children}
    </button>;
};

export default App;
