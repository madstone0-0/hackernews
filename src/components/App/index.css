.page {
    margin: 20px;
}
.interactions {
    text-align: center;
}

.search .form > button.btn.btn-primary {
    margin-bottom: 5px;
}

.dark .search button,
.dark .dark-button {
    background-color: #ff1744;
    border: 1px solid #ff1744;
}

.dark,
.dark html {
    color: #ffffff;
    background: #424242;
    font: 400 16px -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen,
        Ubuntu, Cantarell, "Open Sans", "Helvetica Neue", sans-serif;
    height: 100%;
}

.dark a {
    color: #ffffff;
    text-decoration: none;
}

.dark button:hover {
    color: #f4f4f4;
}

.dark button {
    border-color: #212121;
}

.dark a:hover {
    text-decoration: underline;
}

.dark ul li {
    list-style: none;
    padding: 0;
    margin: 0;
}

.dark input {
    padding: 10px;
    border-radius: 5px;
    outline: none;
    margin-right: 10px;
    border: 1px solid #dddddd;
}

.dark .btn-primary:not(:disabled):not(.disabled).active,
.dark .btn-primary:not(:disabled):not(.disabled):active,
.show > .btn-primary.dropdown-toggle {
    color: #fff;
    background-color: #ff1744;
    border-color: #d32f2f;
}

.dark .btn-primary.focus,
.dark .btn-primary:focus,
.dark .btn-primary:not(:disabled):not(.disabled).active:focus,
.dark .btn-primary:not(:disabled):not(.disabled):active:focus,
.dark .show > .dark .btn-primary.dropdown-toggle:focus {
    box-shadow: 0 0 0 0.2rem rgba(234, 7, 7, 0.5);
}

/* Table styles */
.dark .page .table {
    margin: 20px 0;
}

.dark .page .table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    margin: 10px 0;
    padding: 10px;
    background: #757575;
    border: 1px solid #424242;
}

.dark .page .button-active {
    border-radius: 0;
    border-bottom: 1px solid #ff1744;
}
