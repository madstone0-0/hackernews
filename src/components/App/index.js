/* eslint-disable no-unused-vars,no-console */
/* eslint-disable no-unused-expressions */
import React, { Component } from "react";
import { sortBy } from "lodash";
import PropTypes from "prop-types";
import "./index.css";
import {
    DEFAULT_HPP,
    DEFAULT_PAGE,
    DEFAULT_QUERY,
    PARAM_HPP,
    PARAM_PAGE,
    PARAM_SEARCH,
    PATH_BASE,
    PATH_SEARCH,
    TAG,
} from "../constants";
import Table from "../Table";
import Search from "../Search";
import Button from "../Button";
import Loading from "../Loading";

class App extends Component {
    constructor(props) {
        super(props);

        // initializes state
        this.state = {
            result: null,
            searchKey: "",
            searchTerm: DEFAULT_QUERY,
            error: null,
            isLoading: false,
            sortKey: "NONE",
            isSortReverse: false,
        };

        // Declare functions for global use with the this identifier
        this.needsToSearchTopstories = this.needsToSearchTopstories.bind(this);
        this.setSearchTopstories = this.setSearchTopstories.bind(this);
        this.fetchSearchTopstories = this.fetchSearchTopstories.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
        this.onDismiss = this.onDismiss.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onSort = this.onSort.bind(this);
    }

    onSort(sortKey) {
        const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
        this.setState({ sortKey, isSortReverse });
    }

    needsToSearchTopstories(searchTerm) {
        return !this.state.results[searchTerm];
    }

    // OnSubmit function for search button enables server side searching
    onSearchSubmit(event) {
        const { searchTerm } = this.state;
        this.setState({ searchKey: searchTerm });
        if (this.needsToSearchTopstories(searchTerm)) {
            this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
        }
        event.preventDefault();
    }

    // Set top stories
    setSearchTopstories(result) {
        const { hits, page } = result;
        const { searchKey, results } = this.state;
        const oldHits =
            results && results[searchKey] ? results[searchKey].hits : [];
        const updatedHits = [...oldHits, ...hits];

        this.setState({
            results: {
                results,
                [searchKey]: { hits: updatedHits, page },
            },
            isLoading: false,
        });
    }

    // Fetch the top stories
    fetchSearchTopstories(searchTerm, page = 0) {
        this.setState({ isLoading: true });
        fetch(
            `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${TAG}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`,
        )
            .then(response => response.json())
            .then(result => this.setSearchTopstories(result))
            .catch(error => this.setState({ error }));
    }

    componentDidMount() {
        const { searchTerm } = this.state;
        this.setState({ searchKey: searchTerm });
        this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
    }

    // Dismiss stories
    onDismiss(id) {
        const { searchKey, results } = this.state;
        const { hits, page } = results[searchKey];
        const isNotID = item => item.objectID !== id;
        const updatedHits = hits.filter(isNotID);
        this.setState({
            results: {
                results,
                [searchKey]: { hits: updatedHits, page },
            },
        });
    }

    //Search stories
    onSearchChange(event) {
        this.setState({ searchTerm: event.target.value });
    }

    render() {
        const { searchTerm, results, searchKey, isLoading, sortKey, isSortReverse } = this.state;
        const page =
            (results && results[searchKey] && results[searchKey].page) || 0;
        const list =
            (results && results[searchKey] && results[searchKey].hits) || [];
        return (
            <div className="page">
                <div className="interactions">
                    {/* Passed values from the App component to the Search an table components */}
                    <Search
                        value={searchTerm}
                        onChange={this.onSearchChange}
                        onSubmit={this.onSearchSubmit}
                    >
                        Search
                    </Search>
                </div>
                <Table
                    list={list}
                    onDismiss={this.onDismiss}
                    sortKey={sortKey}
                    onSort={this.onSort}
                    isSortReverse={isSortReverse}
                />
                <div className="interactions">
                    <ButtonWithLoading
                        isLoading={isLoading}
                        onClick={() => this.fetchSearchTopstories(searchKey, page + 1)}
                    >
                        More
                    </ButtonWithLoading>
                </div>
            </div>
        );
    }
}


const withLoading = Component => ({ isLoading, ...rest }) =>
    isLoading ? <Loading /> : <Component {...rest} />;

const ButtonWithLoading = withLoading(Button);

ButtonWithLoading.propTypes = {
    onClick: PropTypes.func.isRequired,
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
};

export default App;
