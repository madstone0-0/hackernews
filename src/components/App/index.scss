.page {
    margin: 20px;
}

.interactions {
    text-align: center;
}

/* Button styles */
.search .form > button.btn.btn-primary {
    margin-bottom: 5px;
}

/* Dark Theme */
.dark, .dark html {
    color: #ffffff;
    background: #424242;
    font: 400 16px -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Open Sans", "Helvetica Neue", sans-serif;
    height: 100%;
}

.dark {

    a {
        color: #ffffff;
        :hover {
            text-decoration: underline;
        }
    }

    ul li {
        list-style: none;
        padding: 0;
        margin: 0;
    }

    input {
        padding: 10px;
        border-radius: 5px;
        outline: none;
        margin-right: 10px;
        border: 1px solid #dddddd;
    }

    .btn-primary:not(:disabled):not(.disabled).active,
    .btn-primary:not(:disabled):not(.disabled):active,
    .show > .btn-primary.dropdown-toggle, .btn-primary.focus, .btn-primary:focus {
        box-shadow: 0 0 0 0.2rem rgba(234, 7, 7, 0.5);
    }

    .btn-primary:not(:disabled):not(.disabled).active, .btn-primary:not(:disabled):not(.disabled):active, .show>.btn-primary.dropdown-toggle {
        background-color: #ff1744;
        border: 1px solid #ff1744;
    }

    .search button,
    .dark-button {
        background-color: #ff1744;
        border: 1px solid #ff1744;
    }

    button:hover {
        color: #f4f4f4;
    }

    button {
        border-color: #212121;
    }

    .page {
        .table {
            margin: 20px 0;
        }

        .table-row {
            display: flex;
            line-height: 24px;
            white-space: nowrap;
            margin: 10px 0;
            padding: 10px;
            background: #757575;
            border: 1px solid #424242;
        }

        .button-active {
            border-radius: 0;
            border-bottom: 1px solid #ff1744;
        }
    }
}
