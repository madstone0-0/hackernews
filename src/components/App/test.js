/* eslint-disable no-unused-vars,no-undef */
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import React from "react";
import App from "./index";

describe("App", function() {
    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<App/>, div);
    });

});

test("snapshots", () => {
    const component = renderer.create(<App/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
