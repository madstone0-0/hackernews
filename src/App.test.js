/* eslint-disable no-undef,no-console */
/* eslint-disable no-unused-vars */
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App from "./App.js";
import { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Search, Button, Table } from "./App.js";
import * as Enzyme from "enzyme";

describe("App", function() {
    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<App/>, div);
    });

});

test("snapshots", () => {
    const component = renderer.create(<App/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});


describe("Search", function() {
    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<Search onChange={() => {
            console.log("Change");
        }} onSubmit={() => {
            console.log("Submitted");
        }}>Search</Search>, div);
    });

});

test("snapshots", () => {
    const component = renderer.create(
        <Search onChange={() => {
            console.log("Change");
        }} onSubmit={() => {
            console.log("Submitted");
        }}> Search </Search>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

describe("Button", () => {
    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<Button onClick={() => {
            console.log("Clicked");
        }}>Give me more daddy</Button>, div);
    });
});

test("snapshots", () => {
    const component = renderer.create(
        <Button onClick={() => {
            console.log("Clicked");
        }}>Give me more daddy</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

const props = {
    list: [
        { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
        { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" }
    ]
};

describe("Table", () => {

    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<Table {...props}/>, div);
    });

    Enzyme.configure({ adapter: new Adapter() });

    it("shows two items in a list", () => {
        const element = shallow(
            <Table {...props} />
        );

        expect(element.find(".table-row").length).toBe(2);
    });
});

test("snapshots", () => {
    const component = renderer.create(
        <Table {...props}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});





