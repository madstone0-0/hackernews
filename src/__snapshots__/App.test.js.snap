// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        <span
          className="glyphicon glyphicon-search"
        />
         
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      More
    </button>
  </div>
</div>
`;

exports[`snapshots 2`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
  />
  <button
    type="submit"
  >
    <span
      className="glyphicon glyphicon-search"
    />
     
     Search 
  </button>
</form>
`;

exports[`snapshots 3`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give me more daddy
</button>
`;

exports[`snapshots 4`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
       comments
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       points
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
       comments
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       points
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
